### PR TITLE
Add saving format limitations and note about masking.

### PR DESCRIPTION
**Pull request by k-w-w**
Mention the masking limitation in the guide (see related issue https://github.com/keras-team/keras/issues/15510)

### CODE REVIEW

File: guides/md/serialization_and_saving.md [hunk 1]
```diff
@@ -292,17 +310,26 @@ reconstructed_model.fit(test_input, test_target)
...
 
-<tensorflow.python.keras.callbacks.History at 0x151d5ac90>
+<keras.callbacks.History at 0x79fb1454b4a8>
 
 ```
 </div>
-#### Limitations
+### Format Limitations
+
+SavedModel limitations:
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:322]: Add line break below

File: guides/md/serialization_and_saving.md [hunk 2]
```diff
@@ -134,9 +141,20 @@ reconstructed_model.fit(test_input, test_target)
...
+2021-11-09 15:11:59.097859: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
+2021-11-09 15:11:59.097888: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (penguin): /proc/driver/nvidia/version does not exist
+2021-11-09 15:11:59.098145: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
+To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
+
+4/4 [==============================] - 0s 2ms/step - loss: 0.3577
 
-<tensorflow.python.keras.callbacks.History at 0x1511b87d0>
+2021-11-09 15:11:59.631545: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:152]: Make sure the md file doesn't include any warnings / log lines this these

File: guides/md/serialization_and_saving.md [hunk 3]
```diff
@@ -292,17 +310,26 @@ reconstructed_model.fit(test_input, test_target)
...
+<keras.callbacks.History at 0x79fb1454b4a8>
 
 ```
 </div>
-#### Limitations
+### Format Limitations
+
+SavedModel limitations:
+- Can be slower and bulkier than H5, since it saves the traced TF graphs of each layer
+- Does not support layers with masks (`layer.supports_masking=True`)
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:324]: This isn't really accurate, is it? Most Keras layers support masking, e.g. Dense, LSTM, etc. If they weren't supported, virtually no model could be saved.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: guides/md/serialization_and_saving.md (hunk 2 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md]: Do we manually remove them after running autogen, or is there a setting that excludes the log lines?

File: guides/md/serialization_and_saving.md (hunk 3 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md]: I reworded this to mention this is the case for custom layers. There is also a note below saying that if the object as a get_config/from_config, then the masking will work.

File: guides/md/serialization_and_saving.md (hunk 1 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md:303]: Done

File: guides/understanding_masking_and_padding.py [hunk 4]
```diff
@@ -264,6 +264,11 @@ def compute_mask(self, inputs, mask=None):
 
 print(mask)
 
+"""
+Note: For more details about format limitations related to masking, see the 
+[serialization guide](/guides/serialization_and_saving).
```
Comment (MarkDaoust) [inline guides/understanding_masking_and_padding.py:269]: I think this link will be broken everywhere except keras.io (github, colab, tensorflow.org, since the paths are all different.

Please use a full URL.

File: guides/md/serialization_and_saving.md [hunk 5]
```diff
@@ -292,17 +291,27 @@ reconstructed_model.fit(test_input, test_target)
...
 
 ```
 </div>
-#### Limitations
+### Format Limitations
+
+SavedModel limitations:
+
+- Can be slower and bulkier than H5, since it saves the traced TF graphs of each layer
+- Can not serialize the mask argument in custom layers (`layer.supports_masking=True`)
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:306]: Rather than `layer.supports_masking=True` this refers to `layer(..., mask=mask_value)`

File: guides/md/serialization_and_saving.md [hunk 6]
```diff
@@ -292,17 +291,27 @@ reconstructed_model.fit(test_input, test_target)
...
+- Does not support models that have a custom training loop.
+  (model overrides `train_step`).
 
-Compared to the SavedModel format, there are two things that don't
-get included in the H5 file:
+Unsupported objects can still be saved and loaded from SavedModel, except they
+must override `get_config`/`from_config`, and the classes must be passed to the
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:311]: `get_config()` / `from_config()`

File: guides/md/serialization_and_saving.md [hunk 7]
```diff
@@ -292,17 +291,27 @@ reconstructed_model.fit(test_input, test_target)
...
 </div>
-#### Limitations
+### Format Limitations
+
+SavedModel limitations:
+
+- Can be slower and bulkier than H5, since it saves the traced TF graphs of each layer
+- Can not serialize the mask argument in custom layers (`layer.supports_masking=True`)
+- Does not support models that have a custom training loop.
+  (model overrides `train_step`).
```
Comment (fchollet) [inline guides/md/serialization_and_saving.md:308]: `train_step()`

**Review by fchollet — COMMENTED** [review]
Thanks for the update

File: guides/understanding_masking_and_padding.py (hunk 4 above)
Comment (k-w-w) [reply guides/understanding_masking_and_padding.py:269]: `/guides/name/` is used in https://github.com/keras-team/keras-io/blob/master/guides/training_with_built_in_methods.py#L27

The guide has the correct links:
https://www.tensorflow.org/guide/keras/train_and_evaluate#introduction
https://keras.io/guides/training_with_built_in_methods/#introduction

File: guides/serialization_and_saving.py [hunk 8]
```diff
@@ -267,10 +267,20 @@ def from_config(cls, config):
...
-get included in the H5 file:
+SavedModel limitations:
+
+- Can be slower and bulkier than H5, since it saves the traced TF graphs of each layer
+- Can not serialize the mask argument in custom layers (`layer.supports_masking=True`)
+- Does not support models that have a custom training loop.
+  (model overrides `train_step`).
+
+Unsupported objects can still be saved and loaded from SavedModel, except they
+must override `get_config`/`from_config`, and the classes must be passed to the
```
[the remark this answers is not in the document]
Comment (k-w-w) [reply guides/serialization_and_saving.py:280]: Changed unsupported to custom, and added a note about the portability.

File: guides/serialization_and_saving.py [hunk 9]
```diff
@@ -267,10 +267,20 @@ def from_config(cls, config):
 
 """
 
-#### Limitations
+### Format Limitations
 
-Compared to the SavedModel format, there are two things that don't
-get included in the H5 file:
+SavedModel limitations:
```
[the remark this answers is not in the document]
Comment (k-w-w) [reply guides/serialization_and_saving.py:272]: This is specific to the Keras SavedModel format (reworded the title)

File: guides/md/serialization_and_saving.md (hunk 5 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md:306]: Changed

File: guides/md/serialization_and_saving.md (hunk 6 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md:311]: done

File: guides/md/serialization_and_saving.md (hunk 7 above)
Comment (k-w-w) [reply guides/md/serialization_and_saving.md:308]: Done

**Review by fchollet — APPROVED** [review]
LGTM, thanks!
